037-prototypes/proposal.md

```diff
@@ -0,0 +1,544 @@
+
+* **Bits**: a directory containing arbitrary data.
+
+  Examples:
+  
+  * source code checked out from a repo
+  * compiled artifacts created by a task or another prototype
+
+* **Resource**: a **prototype** which implements the following messages:
+
+  * `check`: discover objects representing versions, in order
+  * `get`: fetch an object version
+
+
+## Interface Types
+
+```go
+// Object is an object receiving messages and being emitted in response to
+// messages.
+type Object map[string]interface{}
+
+// InfoRequest is the payload written to stdin for the `./info` script.
+type InfoRequest struct {
+  // The properties specifying the object to act on.
+  Object Object `json:"object"`
+
+  // Configuration for handling TLS.
+  TLS TLSConfig `json:"tls,omitempty"`
+}
+
+// InfoResponse is the payload written to stdout from the `./info` script.
+type InfoResponse struct {
+  // The version of the prototype interface that this prototype conforms to.
+  InterfaceVersion string `json:"interface_version"`
+
+  // An optional icon to show to the user.
+  //
+  // Icons must be namespaced by in order to explicitly reference an icon set
+  // supported by Concourse, e.g. 'mdi:' for Material Design Icons.
+  Icon string `json:"icon,omitempty"`
+
+  // The messages supported by the object.
+  Messages []string `json:"messages,omitempty"`
+}
+
+// MessageRequest is the payload written to stdin for a message.
+type MessageRequest struct {
+  // The properties specifying the object to act on.
+  Object Object `json:"object"`
+
+  // Configuration for handling TLS.
+  TLS TLSConfig `json:"tls,omitempty"`
+
+  // Path to a file into which the message handler must write its response.
+  ResponsePath string `json:"response_path"`
+}
+
+// MessageResponse is written to the `response_path` for each object returned
+// by the message. Multiple responses may be written to the same file,
+// concatenated as a JSON stream.
+type MessageResponse struct {
+  // The identifying properties for the object.
+  Object Object `json:"object"`
+
+  // Metadata to associate with the properties. Shown to the user.
+  Metadata []MetadataField `json:"metadata,omitempty"`
+}
+
+// TLSConfig captures common configuration for communicating with servers over
+// TLS.
+type TLSConfig struct {
+  // An array of CA certificates to trust.
+  CACerts []string `json:"ca_certs,omitempty"`
+
+  // Skip certificate verification, effectively making communication insecure.
+  SkipVerification bool `json:"skip_verification,omitempty"`
+}
+
+// MetadataField represents a named bit of metadata associated to an object.
+type MetadataField struct {
+  Name  string `json:"name"`
+  Value string `json:"value"`
+}
+```
+
+
+## Prototype Info
+
+Prior to sending any message, the default command (i.e.
+[`CMD`](https://docs.docker.com/engine/reference/builder/#cmd)) for the image
+will be executed with an `InfoRequest` piped to `stdin`. This request contains
```

Comment:
I wonder if if we broke from the current flow of piping into `stdin` and writing to `stdout`, things would be simpler when it comes to systems where it's easy to put things on the filesystem (and read results from it later) - 👀 k8s runtime.

maybe this is totally crazy, but, could we even argue that perhaps the process boundary is not the right one? (this is from someone who has been reading about https://github.com/fastly/lucet 😅 where sandboxing/"containerizing" happens within the same process)

Reply:
Scratch that, I forgot gRPC actually is language-agnostic. My concern then would be whether prototype authors would be OK with that overhead vs. simple HTTP.

For HTTP, it could work like this:

* Replace Info with a Start request, which spins up a HTTP server for the given object.
* Messages would be handled by invoking their endpoint, i.e. a `foo` message would be something `PUT /messages/foo`.
  * All endpoints would be technically `PUT` since they perform side-effects and should be idempotent.
* Messages write their response over HTTP via their handler, and no longer have to touch the filesystem. We could leverage HTTP chunked responses here.

I kind of like how the object is represented as a server with its own endpoints. In this setup we don't even have to pass the object twice.

One challenge here though is routing traffic to the server. It's on the container's local network, which is unreachable from the `web` nodes. You would need some kind of proxy between them.

Reply:
One complication with an HTTP approach: it seems like it would be difficult to re-attach to a message/action that's in-flight or has finished while the `web` node was away.

With writing the result to a filesystem, we can detach from the process and re-attach when we come back, reading the response from the filesystem if the process has already exited.

With HTTP, the client would disconnect, so we'd still need to figure out where to put the response.

At this point I'm leaning towards just making `Info` write its response to the filesystem too. :thinking: